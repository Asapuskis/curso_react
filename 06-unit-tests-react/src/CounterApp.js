// Los nombres de los componentes deberían ir en UpperCamelCase!

import React, { useState } from 'react';
import PropTypes from 'prop-types';

// Functional Components: https://es.reactjs.org/docs/components-and-props.html
// Eventos en react: https://es.reactjs.org/docs/events.html

const CounterApp = ( { value = 0 } ) => {

    // Usaremos el hook useState para almacenar la variable: https://es.reactjs.org/docs/hooks-state.html
    // useState devuelve un array con el valor definido por parámetro y una función que sirve para modificar ese valor.
    const [ counter, setCounter ] = useState(value);

    // handleAdd sumará 1 al valor de counter
    const handleAdd = (e) => {
        setCounter( (counter) => (counter + 1));
    }

    // handleSubstract restará 1 al valor de counter
    const handleSubstract = (e) => {
        setCounter( (counter) => (counter - 1));
    }

    // handleReset resetea el valor al valor por defecto
    const handleReset = (e) => {
        setCounter(value);
    }

    // Añadimos un botón que va a llamar a la función handleAdd. e es lo que devuelve la función onClick (React.MouseEventHandler<HTMLButtonElement>)
    // No hace falta enviarlo, solo si se van a utilizar alguna de esos valores.
    // return (
    //     <>
    //         <h1>CounterApp</h1>
    //         <h2>{counter}</h2>
    //         <button onClick={ (e) => handleAdd(e) }>+1</button>
    //         <button onClick={ (e) => handleReset(e) }>Reset</button>
    //         <button onClick={ (e) => handleSubstract(e) }>-1</button>
    //     </>
    // );

    // Este return es igual que el anterior pero no envía los atributos del evento e ya que no hacen falta
    return (
        <>
            <h1>CounterApp</h1>
            <h2>{counter}</h2>
            <button onClick={  handleAdd }>+1</button>
            <button onClick={ handleReset }>Reset</button>
            <button onClick={ handleSubstract }>-1</button>
        </>
    );
}

CounterApp.propTypes = {
    value: PropTypes.number.isRequired
}

export default CounterApp;