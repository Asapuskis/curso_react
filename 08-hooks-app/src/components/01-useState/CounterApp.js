import React, { useState } from 'react';
import './counter.css';

export const CounterApp = () => {

    const counterState = {
        counter1: 0,
        counter2: 20,
        counter3: 30
    };

    // Podemos definir un state como un objeto, no tiene que ser una variable primitiva por defecto.
    const [ counter, setCounter] = useState(counterState);

    // Para acceder a los atributos del objeto, desestructuramos y cogemos los que nos interesan.
    const { counter1, counter2 } = counter;

    // En este caso, solo queremos aumentar uno de los contadores, el counter1. Si pusiésemos solo counter1 + 1, reemplazaría el objeto por counter1
    // Por lo tanto, para mantener el estado anterior utilizamos el operador spread ... y además, sustituímos counter1 por counter1 + 1
    const handleClick = () => {
        setCounter({
            ...counter,
            counter1: counter1 + 1
        });
    }

    return (
        <>
            <h1>Counter { counter1 }</h1>
            <h1>Counter { counter2 }</h1>
            <hr/>
            <button className='btn btn-primary' onClick={handleClick}>
                +1
            </button>
        </>
    )
};
