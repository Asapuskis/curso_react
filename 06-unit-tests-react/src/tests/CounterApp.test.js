import { shallow } from 'enzyme';
import CounterApp from "../CounterApp";


describe('Pruebas en <CounterApp />', () => {

    test('Debe mostrar <CounterApp /> correctamente', () => {
        // Utilizaremos shallow para garantizar recoger lo que se rendereiza.
        const wrapper = shallow( <CounterApp value={0} /> );

        // toMatchSnapshot genera un archivo .snap. Estos archivos son fotografías de como queda el componente.
        // Previamente hay que instalar enzyme-to-json: npm install --save-dev enzyme-to-json
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe mostrar <CounterApp /> con valor 100', () => {
        // Utilizaremos shallow para garantizar recoger lo que se rendereiza.
        const wrapper = shallow( <CounterApp value={100}/> );

        // toMatchSnapshot genera un archivo .snap. Estos archivos son fotografías de como queda el componente.
        // Previamente hay que instalar enzyme-to-json: npm install --save-dev enzyme-to-json
        expect( wrapper ).toMatchSnapshot();
        const textoh2 = wrapper.find('h2').text().trim();
        console.log(textoh2);
        expect(textoh2).toBe('100');
    });

    test('Debe incrementar el contador en 1', () => {
        const wrapper = shallow( <CounterApp value={0}/> );
        //Se pueden simular eventos con la función simulate y añadiendo el tipo de evento que queremos simular.
        const btnAdd = wrapper.find('button').at(0).simulate('click');
        console.log(btnAdd.html());
        const textoh2 = wrapper.find('h2').text().trim();
        console.log(textoh2);
        expect(textoh2).toBe('1');
    });

    test('Debe decrementar el contador en 1', () => {
        const wrapper = shallow( <CounterApp value={10}/> );
        //Se pueden simular eventos con la función simulate y añadiendo el tipo de evento que queremos simular.
        const btnSubstract = wrapper.find('button').at(2).simulate('click');
        console.log(btnSubstract.html());
        const textoh2 = wrapper.find('h2').text().trim();
        console.log(textoh2);
        expect(textoh2).toBe('9');
    });

    test('Debe resetear el contador en 1', () => {
        const wrapper = shallow( <CounterApp value={105}/> );

        wrapper.find('button').at(0).simulate('click');
        let textoh2 = wrapper.find('h2').text().trim();
        console.log(textoh2);

        wrapper.find('button').at(1).simulate('click');
        textoh2 = wrapper.find('h2').text().trim();
        console.log(textoh2);
        expect(textoh2).toBe('105');
    });

});