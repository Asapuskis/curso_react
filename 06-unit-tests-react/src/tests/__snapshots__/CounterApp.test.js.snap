// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CounterApp /> Debe mostrar <CounterApp /> con valor 100 1`] = `
<Fragment>
  <h1>
    CounterApp
  </h1>
  <h2>
    100
  </h2>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
</Fragment>
`;

exports[`Pruebas en <CounterApp /> Debe mostrar <CounterApp /> correctamente 1`] = `
<Fragment>
  <h1>
    CounterApp
  </h1>
  <h2>
    0
  </h2>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
</Fragment>
`;
